app/ante/authz_test.go

```diff
@@ -0,0 +1,550 @@
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/types/tx/signing"
+	authsigning "github.com/cosmos/cosmos-sdk/x/auth/signing"
+	sdkvesting "github.com/cosmos/cosmos-sdk/x/auth/vesting/types"
+	"github.com/cosmos/cosmos-sdk/x/authz"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+	"github.com/evmos/ethermint/testutil"
+	utiltx "github.com/evmos/ethermint/testutil/tx"
+
+	"github.com/evmos/ethermint/app/ante"
+
+	"github.com/evmos/ethermint/app"
+	"github.com/evmos/ethermint/crypto/ethsecp256k1"
+	"github.com/evmos/ethermint/encoding"
+
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+	evmtypes "github.com/evmos/ethermint/x/evm/types"
+)
+
+func generatePrivKeyAddressPairs(accCount int) ([]*ethsecp256k1.PrivKey, []sdk.AccAddress, error) {
+	var (
+		err           error
+		testPrivKeys  = make([]*ethsecp256k1.PrivKey, accCount)
+		testAddresses = make([]sdk.AccAddress, accCount)
+	)
+
+	for i := range testPrivKeys {
+		testPrivKeys[i], err = ethsecp256k1.GenerateKey()
+		if err != nil {
+			return nil, nil, err
+		}
+		testAddresses[i] = testPrivKeys[i].PubKey().Address().Bytes()
+	}
+	return testPrivKeys, testAddresses, nil
+}
+
+func newMsgExec(grantee sdk.AccAddress, msgs []sdk.Msg) *authz.MsgExec {
+	msg := authz.NewMsgExec(grantee, msgs)
+	return &msg
+}
+
+func newMsgGrant(granter sdk.AccAddress, grantee sdk.AccAddress, a authz.Authorization, expiration *time.Time) *authz.MsgGrant {
+	msg, err := authz.NewMsgGrant(granter, grantee, a, expiration)
+	if err != nil {
+		panic(err)
+	}
+	return msg
+}
+
+func createNestedMsgExec(a sdk.AccAddress, nestedLvl int, lastLvlMsgs []sdk.Msg) *authz.MsgExec {
+	msgs := make([]*authz.MsgExec, nestedLvl)
+	for i := range msgs {
+		if i == 0 {
+			msgs[i] = newMsgExec(a, lastLvlMsgs)
+			continue
+		}
+		msgs[i] = newMsgExec(a, []sdk.Msg{msgs[i-1]})
+	}
+	return msgs[nestedLvl-1]
+}
+
+func TestAuthzLimiterDecorator(t *testing.T) {
+	testPrivKeys, testAddresses, err := generatePrivKeyAddressPairs(5)
+	require.NoError(t, err)
+
+	distantFuture := time.Date(9000, 1, 1, 0, 0, 0, 0, time.UTC)
+
+	validator := sdk.ValAddress(testAddresses[4])
+	stakingAuthDelegate, err := stakingtypes.NewStakeAuthorization([]sdk.ValAddress{validator}, nil, stakingtypes.AuthorizationType_AUTHORIZATION_TYPE_DELEGATE, nil)
+	require.NoError(t, err)
+
+	stakingAuthUndelegate, err := stakingtypes.NewStakeAuthorization([]sdk.ValAddress{validator}, nil, stakingtypes.AuthorizationType_AUTHORIZATION_TYPE_UNDELEGATE, nil)
+	require.NoError(t, err)
+
+	decorator := ante.NewAuthzLimiterDecorator(
+		sdk.MsgTypeURL(&evmtypes.MsgEthereumTx{}),
+		sdk.MsgTypeURL(&stakingtypes.MsgUndelegate{}),
+	)
+
+	testCases := []struct {
+		name        string
+		msgs        []sdk.Msg
+		checkTx     bool
+		expectedErr error
+	}{
+		{
+			"enabled msg - non blocked msg",
+			[]sdk.Msg{
+				banktypes.NewMsgSend(
+					testAddresses[0],
```

Comment:
nit: can move `testAddresses` to file-global variables, then use them from the helper methods, to avoid passing them each time.